internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixReadOnlyStoreConfigRepository.java

```diff
@@ -131,62 +130,70 @@ public StoreConfig getStoreConfigOrThrow(String storeName) {
-        Map<String, StoreConfig> map = new HashMap<>(storeConfigMap.get());
-        List<String> newStores =
-            currentChildren.stream().filter(newStore -> !map.containsKey(newStore)).collect(Collectors.toList());
-
-        Set<String> deletedStores = new HashSet<>(map.keySet());
-        currentChildren.forEach(deletedStores::remove);
+    public void handleChildChange(String parentPath, List<String> currentChildren) {
+      synchronized (availableStoreSet) {
+        Set<String> storeSetSnapshot = new HashSet<>(availableStoreSet.get());
+        List<String> newStores = currentChildren.stream()
+            .filter(newStore -> !storeSetSnapshot.contains(newStore))
+            .collect(Collectors.toList());
+
+        // obtain the stores that are removed
+        currentChildren.forEach(storeSetSnapshot::remove);
         LOGGER.info(
             "Store configs list is changed. {} new configs. And will delete {} configs.",
             newStores.size(),
-            deletedStores.size());
-        // New added store configs
-        List<StoreConfig> newConfigs = accessor.getStoreConfigs(newStores);
-        for (StoreConfig config: newConfigs) {
-          map.put(config.getStoreName(), config);
-          accessor.subscribeStoreConfigDataChangedListener(config.getStoreName(), storeConfigChangedListener);
-        }
-
-        // Deleted store configs
-        for (String deletedStore: deletedStores) {
-          map.remove(deletedStore);
-          accessor.unsubscribeStoreConfigDataChangedListener(deletedStore, storeConfigChangedListener);
+            storeSetSnapshot.size());
+
+        // update the available store set
+        availableStoreSet.set(new HashSet<>(currentChildren));
+
+        synchronized (loadedStoreConfigMap) {
+          Map<String, StoreConfig> map = new HashMap<>(loadedStoreConfigMap.get());
+          // Deleted store configs
+          for (String deletedStore: storeSetSnapshot) {
+            map.remove(deletedStore);
+            accessor.unsubscribeStoreConfigDataChangedListener(deletedStore, storeConfigChangedListener);
+          }
+          loadedStoreConfigMap.set(map);
         }
-        storeConfigMap.set(map);
       }
     }
   }
 
   protected class StoreConfigChangedListener implements IZkDataListener {
     @Override
-    public void handleDataChange(String dataPath, Object data) throws Exception {
+    public void handleDataChange(String dataPath, Object data) {
       if (!(data instanceof StoreConfig)) {
         throw new VeniceException(
             "Invalid data from zk notification. Required: StoreConfig, but get: " + data.getClass().getName());
       }
       StoreConfig config = (StoreConfig) data;
-      synchronized (storeConfigMap) {
-        Map<String, StoreConfig> map = new HashMap<>(storeConfigMap.get());
+      synchronized (loadedStoreConfigMap) {
+        Map<String, StoreConfig> map = new HashMap<>(loadedStoreConfigMap.get());
```

Comment:
This is not quite right.
If the `loadedStoreConfigMap` is using concurrent hashmap, we don't need this section, instead we can update the map directly, right?

Reply:
Double-checked the codes. There's no need to swap the entire map atomically so I will remove the AtomicReference and just use concurrent hashmap